CRM/Activityreport/Data.php

```diff
@@ -50,12 +161,63 @@ public static function get($offset = 0, $limit = 0, $multiValuesOffset = 0, $sta
       $params['activity_date_time'] = $activityDateFilter;
     }
 
-    $activities = civicrm_api3('Activity', 'get', $params);
+    while ($offset < $total) {
+      if ($offset) {
+        $offset--;
+      }
 
-    $formattedActivities = self::formatResult($activities['values']);
-    $result = self::splitMultiValues($formattedActivities, $offset, $multiValuesOffset);
+      $params['options']['offset'] = $offset;
 
-    return $result;
+      $activities = civicrm_api3('Activity', 'get', $params);
+
+      $formattedActivities = self::formatResult($activities['values']);
+
+      unset($activities);
+
+      while (!empty($formattedActivities)) {
+        $result = self::splitMultiValues($formattedActivities, $offset, $multiValuesOffset);
+
+        if ($result['info']['date'] !== $date) {
+          $page = 0;
+          $date = $result['info']['date'];
+        }
+
+        $cachedCount += self::cachePacket($result['data'], $date, $page++);
+
+        unset($result['data']);
+
+        $formattedActivities = array_slice($formattedActivities, $result['info']['nextOffset'] - $offset, NULL, TRUE);
+
+        $offset = $result['info']['nextOffset'];
+        $multiValuesOffset =  $result['info']['multiValuesOffset'];
+      }
+    }
+
+    return $cachedCount;
```

Comment:
@coldrunKacper `processAllActivities` doesn't sound good to me. It's just too generic and it can mean anything. Maybe `getPaginatedResults` or `getPaginatedActivities` would be more specific? 

Also, as I mentioned before, this method is doing 3 things. I understand that maybe it will be hard to separate "fetch data" from "split it into pages", but caching should definitely be dealt somewhere else. This class shouldn't know anything about caching at all!

Ideally, what we want is two separate things:
1. Being able to build a report from data loaded from the DB
2. Being able to cache it somewhere

And then we have a third party here that will glue these two things together, loading the data using `#1` and then passing this data to `#2`. If we have well defined interfaces for this, we should be able to use `#2` to cache any kinda of report provided by `#1`, as long as it implement this interface. The way it's implemented right now, there are almost no interfaces. There's this single class load loads the data and caches itself. If we were to implement a new report, we would probably have to duplicate some of the code from this class.

With that said, I agree we have enough to test the performance improvements here and check if this paginated approach is working.